Fails to build with GCC 15 on ppc64le due to removed intrinsics

**Describe the bug**
We’re now testing a pre-release of GCC 15 in Fedora Rawhide (the development version of the distribution). Sleef fails to build on `ppc64le` with GCC 15 because it uses two intrinsics that were removed as redundant in GCC 15:

- `__builtin_vsx_xvcvsxwdp()`, removed in https://github.com/gcc-mirror/gcc/commit/fd9fdb33ae252ec34cc33675433eb56637905257
- `__builtin_vsx_xvcvdpsxws()`, removed in https://github.com/gcc-mirror/gcc/commit/224cc560a6ac19c9454038efe6230096b46f4806

Each GCC commit message linked above provides a recommendation for what to use instead. Both of the removed builtins affecting Sleef were undocumented and had no test cases.

**Command lines and logs**

I’m not going to put much effort into this section because the cause of the issue is so clear. I’m attaching a [log from a failing build](https://github.com/user-attachments/files/18495860/build.log), and excerpting the error below. I can provide more information if it turns out to be necessary.

```
[45/352] /usr/bin/gcc -DDORENAME=1 -DENABLEFLOAT128=1 -DENABLE_ALIAS=1 -DENABLE_BUILTIN_MATH=1 -DENABLE_SYS_getrandom=1 -DENABLE_VSX3=1 -I/builddir/build/BUILD/sleef-3.7.0-build/sleef-3.7/src/common -I/builddir/build/BUILD/sleef-3.7.0-build/sleef-3.7/src/arch -I/builddir/build/BUILD/sleef-3.7.0-build/sleef-3.7/redhat-linux-build/src/quad/include -I/builddir/build/BUILD/sleef-3.7.0-build/sleef-3.7/src/quad -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64 -mcpu=power8 -mtune=power8 -fasynchronous-unwind-tables -fstack-clash-protection -Wall -Wno-unused-function -Wno-attributes -Wno-unused-result -Wno-psabi -ffp-contract=off -fno-math-errno -fno-trapping-math -DNDEBUG -std=gnu99 -fPIC -mcpu=power9 -MD -MT src/quad/CMakeFiles/sleefquadvsx3_obj.dir/sleefsimdqp.c.o -MF src/quad/CMakeFiles/sleefquadvsx3_obj.dir/sleefsimdqp.c.o.d -o src/quad/CMakeFiles/sleefquadvsx3_obj.dir/sleefsimdqp.c.o -c /builddir/build/BUILD/sleef-3.7.0-build/sleef-3.7/src/quad/sleefsimdqp.c
FAILED: src/quad/CMakeFiles/sleefquadvsx3_obj.dir/sleefsimdqp.c.o 
/usr/bin/gcc -DDORENAME=1 -DENABLEFLOAT128=1 -DENABLE_ALIAS=1 -DENABLE_BUILTIN_MATH=1 -DENABLE_SYS_getrandom=1 -DENABLE_VSX3=1 -I/builddir/build/BUILD/sleef-3.7.0-build/sleef-3.7/src/common -I/builddir/build/BUILD/sleef-3.7.0-build/sleef-3.7/src/arch -I/builddir/build/BUILD/sleef-3.7.0-build/sleef-3.7/redhat-linux-build/src/quad/include -I/builddir/build/BUILD/sleef-3.7.0-build/sleef-3.7/src/quad -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64 -mcpu=power8 -mtune=power8 -fasynchronous-unwind-tables -fstack-clash-protection -Wall -Wno-unused-function -Wno-attributes -Wno-unused-result -Wno-psabi -ffp-contract=off -fno-math-errno -fno-trapping-math -DNDEBUG -std=gnu99 -fPIC -mcpu=power9 -MD -MT src/quad/CMakeFiles/sleefquadvsx3_obj.dir/sleefsimdqp.c.o -MF src/quad/CMakeFiles/sleefquadvsx3_obj.dir/sleefsimdqp.c.o.d -o src/quad/CMakeFiles/sleefquadvsx3_obj.dir/sleefsimdqp.c.o -c /builddir/build/BUILD/sleef-3.7.0-build/sleef-3.7/src/quad/sleefsimdqp.c
In file included from /builddir/build/BUILD/sleef-3.7.0-build/sleef-3.7/src/quad/sleefsimdqp.c:124:
/builddir/build/BUILD/sleef-3.7.0-build/sleef-3.7/src/arch/helperpower_128.h: In function ‘vcast_vd_vi’:
/builddir/build/BUILD/sleef-3.7.0-build/sleef-3.7/src/arch/helperpower_128.h:375:9: error: implicit declaration of function ‘__builtin_vsx_xvcvsxwdp’; did you mean ‘__builtin_vsx_xvcvsxddp’? [-Wimplicit-function-declaration]
  375 |   ret = __builtin_vsx_xvcvsxwdp(swap);
      |         ^~~~~~~~~~~~~~~~~~~~~~~
      |         __builtin_vsx_xvcvsxddp
/builddir/build/BUILD/sleef-3.7.0-build/sleef-3.7/src/arch/helperpower_128.h:375:3: error: AltiVec argument passed to unprototyped function
  375 |   ret = __builtin_vsx_xvcvsxwdp(swap);
      |   ^~~
/builddir/build/BUILD/sleef-3.7.0-build/sleef-3.7/src/arch/helperpower_128.h: In function ‘vtruncate_vi_vd’:
/builddir/build/BUILD/sleef-3.7.0-build/sleef-3.7/src/arch/helperpower_128.h:410:9: error: implicit declaration of function ‘__builtin_vsx_xvcvdpsxws’; did you mean ‘__builtin_vsx_xvcvdpsxds’? [-Wimplicit-function-declaration]
  410 |   ret = __builtin_vsx_xvcvdpsxws(vd);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~
      |         __builtin_vsx_xvcvdpsxds
/builddir/build/BUILD/sleef-3.7.0-build/sleef-3.7/src/arch/helperpower_128.h:410:3: error: AltiVec argument passed to unprototyped function
  410 |   ret = __builtin_vsx_xvcvdpsxws(vd);
      |   ^~~
```

I found the second removed builtin incrementally, by working around the first error and doing another test build.

**To Reproduce**
Just build with GCC 15. It may or may not be necessary to set `-DSLEEF_BUILD_QUAD:BOOL=TRUE`, since I saw the error while compiling the quad library.

**Expected behavior**
Sleef compiles successfully, and tests pass, as it does on GCC 14.

**Screenshots**
N/A

**Environment**
 - Is it a native build or cross-compilation? **native**
 - Host/Target Architecture [e.g. x86_64 if native build, x86_64/aarch64 if cross build, ..] `ppc64le`
 - OS with version [e.g. Linux/Ubuntu-22.04, Windows, macOS, iOS, ...] **Fedora Rawhide**
 - CMake version [e.g. 3.18, ...] **3.31.4**
 - Makefile Generator [e.g. GNU Make, Ninja, ...] **Ninja**
 - Compiler with version [e.g. gcc-14, clang-18, apple-clang-15.2, msvc-15.0, ...] **gcc-15 (15.0.1 20250114 (Red Hat 15.0.1-0))**
 - If CUDA, please provide the version of the CUDA ToolKit. **N/A**

**Additional context**

The call to each of these intrinsics is guarded by

```
#if defined(__clang__) || __GNUC__ >= 7
```

with a fallback implementation using inline assembly.

A quick fix, then, is to replace each such conditional with

```
#if defined(__clang__) || (__GNUC__ >= 7 && __GNUC__ < 15)
```

and indeed this is working as an interim patch in Fedora. I’ll open a PR with this as a candidate quick fix.

It might be even better to follow the advice of the individual GCC commit messages, e.g. replace `__builtin_vsx_xvcvsxwdp()` with `vec_doublee` or `vec_doubleo`, but given my lack of familiarity with the POWER ISA and available time, I think this is beyond what I’ll be able to contribute. However, I would be happy to help test such an approach.

--- a/src/arch/helperpower_128.h
+++ b/src/arch/helperpower_128.h
@@ -371,7 +371,7 @@ static INLINE vdouble vcast_vd_vi(vint vi)
 {
   vdouble ret;
   vint swap = vec_mergeh(vi, vi);
-#if defined(__clang__) || __GNUC__ >= 7
+#if defined(__clang__) || (__GNUC__ >= 7 && __GNUC__ < 15)
   ret = __builtin_vsx_xvcvsxwdp(swap);
 #else
   __asm__ __volatile__("xvcvsxwdp %x0,%x1" : "=wa" (ret) : "wa" (swap));
@@ -406,7 +406,7 @@ static INLINE vint2 vtruncate_vi2_vf(vfloat vf)
 static INLINE vint vtruncate_vi_vd(vdouble vd)
 {
   vint ret;
-#if defined(__clang__) || __GNUC__ >= 7
+#if defined(__clang__) || (__GNUC__ >= 7 && __GNUC__ < 15)
   ret = __builtin_vsx_xvcvdpsxws(vd);
 #else
   __asm__ __volatile__("xvcvdpsxws %x0,%x1" : "=wa" (ret) : "wa" (vd));

